File: provider/src/main/java/com/ibashkimi/provider/filter/KalmanFilter.java
package com.ibashkimi.provider.filter;

public class KalmanFilter implements Filter {
    private double q; //process noise covariance
    private double r; //measurement noise covariance
    private double x; //value
    private double p; //estimation error covariance
    private double k; //kalman gain

    public KalmanFilter() {
        this(0, 0, 0, 0);
    }

    public KalmanFilter(double q, double r, double x, double p) {
        this.q = q;
        this.r = r;
        this.x = x;
        this.p = p;
    }


    @Override
    public double doJob(double input) {
        p = p + q;
        k = p / (p + r);
        x = x + k * (input - x);
        p = (1 - k) * p;
        return x;
    }
}

File: provider/src/main/java/com/ibashkimi/provider/filter/KalmanFilter.kt
package com.ibashkimi.provider.filter

class KalmanFilter(
    private val q: Double = 0.0, //process noise covariance
    private val r: Double = 0.0, //measurement noise covariance
    private var x: Double = 0.0, //value
    private var p: Double = 0.0  // estimation error covariance
) : Filter {

    private var k: Double = 0.0 //kalman gain

    override fun doJob(input: Double): Double {
        p += q
        k = p / (p + r)
        x += k * (input - x)
        p *= (1 - k)
        return x
    }

}